### ISSUE TITLE
New general fixed-step integrator method

### ISSUE THREAD

**Issue by gaffarelj** [tudat-team/tudat#81]
The Euler and fixed-step RK4 integrators should be re-made into a new fixed-step integrator method that takes (amongst others) a Butcher tableau as input.

The Euler and RK4 integrators can then be replaced by this method with the corresponding tableau.

Because this new method should be generic, different tableaus (taken from a new or already implemented enum/switch) allows for a more straightforward implementation of fixed-step RK5, 6, 7, 8, etc integrators (instead of using variable step integrators with tolerances of inf).

**Comment by gaffarelj** [comment]
Work for this is carried in the [feature/fixed_step_integrators](https://github.com/tudat-team/tudat/tree/feature/fixed_step_integrators) branch.

**Comment by gaffarelj** [comment]
Good afternoon @DominicDirkx and @geoffreygarrett (since you also wanted to stay in the loop if I remember correctly).

I wanted to update you on what I've done so far for the integrators refactoring, to make sure I'm still going in the right direction.
I'm working in the `feature/fixed_step_integrators` branch of tudat and tudatpy.

What I have done (and manually tested in C++ and Python):

1.  I have added a function in `rungeKuttaCoefficients.cpp` to print the Butcher tableau of a given coefficient step. This has been exposed to Python as propagation_setup.integrator.print_butcher_tableau. For example, one can run `propagation_setup.integrator.print_butcher_tableau(propagation_setup.integrator.RKCoefficientSets.rkf_45)`.

2.  I have added a Euler and RK4 coefficient set in `rungeKuttaCoefficients.cpp`. They have also been exposed to Python (so their Butcher tableau can also be printed using the new function above).
3.  I have added a new class in the integrators, in `rungeKuttaFixedStepSizeIntegrator.h`. It contains all of the method a Tudat integrator should have, and takes a `RungeKuttaCoefficients::CoefficientSets` as one of its input. This class can thus be used (using the appropriate Butcher tableau) as a new fixed step Euler or fixed step RK4.

What I plan on doing next:

4.  Make the new `rungeKuttaFixedStepSizeIntegrator` also accept what are currently variable step coefficient sets (with two rows in the "b" column of their Butcher tableau). I will thus add an additional input to `rungeKuttaFixedStepSizeIntegrator` to specify which order to use for the fixed step integrator. This way, users could specify that they want to use for instance RKF_78 as a fixed step coefficient set, using only the 8th order.
5.  I think I should take some time to add the new fixed step integrator class into `integrators/createNumericalIntegrator.h`, so user/methods that use this interface will recognise the new fixed step integrator.
6.  I want to replace the current `euler.h` and `rungeKutte4Integrator.h` by instances of the new `rungeKuttaFixedStepSizeIntegrator`, with the Butcher tableau input pre-set to the Euler and RK4 coefficient sets.
7. There are quite a few fixed step coefficient sets that could (I think) very easily be added, for instance from [this list](https://en.wikipedia.org/wiki/List_of_Runge%E2%80%93Kutta_methods#Explicit_methods).
8. Writing unit tests for all of these (the current ones still pass so I didn't break anything so far).

**Comment by DominicDirkx** [comment]
Hi Jeremie, this all looks really nice, something like this has been on my list of things to do for a long time, good to see it coming to fruition! Below a few comments/suggestions for the coming steps:

- The CoefficientSets enum is now a member of the RungeKuttaCoefficients. This is kind of a weird design (not yours, but one we made very long ago). I propose to move this into the tudat::numerical_integrators namespace
- Right now here: https://github.com/tudat-team/tudat/blob/171368d27e48a9703deb742cd73de113abb486b3/src/math/integrators/rungeKuttaCoefficients.cpp#L56-L57 (and places like it) a coefficient set for a fixed-step integrator is still defined as having a 'lower' and 'higher' order. I suggest we remove this, and don't define this if the coefficient set is for a fixed-step integrator. This will probably require some changes to the RungeKuttaCoefficients class, and its usages. It's not critical, and we can discuss how to resolve this sometime in the coming weeks.
- Writing proper unit tests of all new integrators (with new Butcher Tableaus) will be difficult. At the very least, we could check the order of the integrators, and for some of them, we can find the leading error term in literature (which we could use if we take a differential equation with known solution as test case). 
- Adding functionality for users to read/write Butcher tableaus to files (and/or matrices?) would be useful. But, this can be done after this work is finished.
- If you're not very familiar with how everything in the createNumericalIntegrator.h is set up, we can have a quick chat about it. Since this is very close to where the user will interact with the code, I think it's the most important part in terms of how consistent everything is with the rest of Tudat, and how easy it is to use

**Comment by gaffarelj** [comment]
Good afternoon,
New update on the progress/plan so far.

### What's new for tudat:

1. The fixed-step coefficients are now defined by a `isFixedStepSize` boolean, instead of setting the higher and lower order to equal values (done with 3343d0c6b5464429cc01491e4b704c2f8e25a323).
2. The coefficients that were defined for variable step integration (e.g. RKF_78) can now be used with the new fixed step RK integrator, if the `orderToUse` (e.g. 7 or 8) is specified at the same time. I added a bunch of checks to make sure that using variable step coefficients for fixed step integration is impossible if the correct order is not specified, and that using fixed step coefficients for variable step integration raises an error (with e78af060906ec5ad852c6157201339f63dda89ac and 343d77eb8249a210992094fdd529a41a55e9b11f).
3. The `CoefficientSets` enum is now placed out of the `RungeKuttaCoefficients` struct, and directly in the `numerical_integrators` namesapce. I also changed everywhere in the code where `RungeKuttaCoefficients::CoefficientSets` was used as a type, to only keep `CoefficientSets`. This change then breaks backward-compatibility (I think). This was pushed with a336815bb697d9402fffc660a0b4c3c3e511a584.
4. Related to the previous point: I considered at the same time splitting the `CoefficientSets` enum into two new enums: `FixedStepCoefficientSets` and `VariableStepCoefficientSets` (or something similar). I decided not to, because I think it would make the code too messy, and having the checks that I implemented seems already clear in my opinion.
5. New methods have been added to `createNumericalIntegrator.h` to create a fixed step RK integrator (with 4f8c275009ba874b3f13ca80249ecaceea1a90b2).
6. New explicit fixed-step coefficient sets have been added (8 to be precise), with ec3d6ec41b1b4e1d48c02e52b40fee7508a81bdd.
7. I have remove the hard-coded euler and RK4 integrators, and replaced them with classes inherited from `rungeKuttaFixedStepSize`, with the euler or RK4 coefficient set used. Both of these integrators should still be usable in C++ and Python as if nothing happened. This was pushed with 37bffb0dcda38d52eaaa4c13359bb38a4275aa88. We may chose to hide these from the API since they can be replace by the new fixed step integrator with the appropriate coefficient set.

### What's new to tudatpy
(all exposed in [expose_integrator_setup.cpp](https://github.com/tudat-team/tudatpy/blob/feature/fixed_step_integrators/tudatpy/kernel/expose_numerical_simulation/expose_propagation_setup/expose_integrator_setup.cpp)):

1. The following use of the fixed step RK integrator is possible (for any fixed step coefficient set):
  ```
  coefficients = propagation_setup.integrator.explicit_trapezoid_rule
  propagation_setup.integrator.print_butcher_tableau(coefficients)
  integrator_settings = propagation_setup.integrator.runge_kutta_fixed_step_size(
      simulation_start_epoch, fixed_step_size, coefficients
  )
  ```
2. The following use of the fixed step RK integrator is possible (for any variable step coefficient set):
  ```
  coefficients = propagation_setup.integrator.rkf_78
  integrator_settings = propagation_setup.integrator.runge_kutta_fixed_step_size(
      simulation_start_epoch, fixed_step_size, coefficients, order_to_use=7
  )
  ```
The above corresponds to the trick of using infinite tolerances (except that no unused order will be computed now):
  ```
  integrator_settings = propagation_setup.integrator.runge_kutta_variable_step_size(
      simulation_start_epoch,
      fixed_step_size,
      coefficients,
      fixed_step_size,
      fixed_step_size,
      relative_error_tolerance=np.inf,
      absolute_error_tolerance=np.inf)
  ```

### Next on the todo list

1. Writing unit tests for the new fixed step integrator and new coefficient sets. Both will be tricky; I manually tested all the new functionalities in C++ and Python and everything looks okay, but I do want to implement more rigurous unit tests.
2. Write documentation: Docstrings (should go fast); add a fixed step integration guide to tudat-space; edit the variable step integration guide on tudat-space (remove discussion about forcing fixed step integration with infinite tolerance, and use the new fixed step RK integator instead, specifying the order to use).
3. Adding a functionality to let users define their own Butcher tableau from 3 matrixes (a, b, c coefficients).

**Comment by DominicDirkx** [comment]
Hi Jeremie, this is progressing really well! Looks like you are close to the final version, with the main open points documentation and testing. Below a small points:

Tudat:

1. Thanks :)
2. why do you use the order as input, instead of lower/higher? If you have lower/higher, there is no chance of it going wrong, and it is then possible to iterate over all variable step tableaus without making sure the input order is correct.
3. Indeed, this will break backwards compatibility, but only in Tudat (?)
4. Sounds like a perfectly good choice
5. Nice
6. Cool! We're going to have to limit what the P&O students use, or their plots will get too full :) Also, with multiple methods of the same order implemented, we're going to have to go into some more detail on error estimates (from theoretical point of view) and the concept of integrator stability to explain why there are so many different ones! I think this will make a nice addition to the course, and fill a gap that has been there until now
7. Nice

Tudatpy
- Looks good, no comment :) 

Todo:
- Unit tests. one thing you can test is the order of the integrator. Pick a relatively simple problem (does not even need to be an astrodynamics problem, but I would recommend a vector problem), so long as the Taylor series expansion of the analytical solution has non-zero values for each order of t. Vary the step size successively (1, 2, 4, 8, 16 seconds) to compute the order of the method. Check this against what they should be. For some of the methods, you may be able to find an implementation of the method in another toolbox. But, for now, I suggest you stick to the test of the order, and we can see where we take it from there. Did you have any specific plans here?

**Comment by DominicDirkx** [comment]
One other request: since you're going all out with adding new methods, how about adding a few really high order ones:

https://sce.uhcl.edu/rungekutta/rk108.txt
https://sce.uhcl.edu/rungekutta/rk1210.txt
https://sce.uhcl.edu/rungekutta/rk1412.txt

On these links you can find entries for the Butcher tableaus of RK10(8), RK12(10 and RK 14(12) methods. I think with some automated file reading/writing magic you could turn these into new entries for tudat.

The literature on these methods (for reference purposes in the docs) can be found here: http://www.dynamicpublishers.com/Neural/NPSC2012/NPSC-20-3-4/14-NPSC-32-SA-25.pdf

**Comment by DominicDirkx** [comment]
I now made the mistake of going down the integrator literatue rabbit hole. A few other options to implement:
https://www.sciencedirect.com/science/article/pii/S0898122111004706

Here, a bunch of others are provided:
https://www.sfu.ca/~jverner/
these have the attractive property of allowing interpolation between stages for dense output (I think). We could implement these if/when we go for dense output

**Comment by gaffarelj** [comment]
Hi Dominic,
Thanks a lot for all your comments!

> 2. why do you use the order as input, instead of lower/higher? If you have lower/higher, there is no chance of it going wrong, and it is then possible to iterate over all variable step tableaus without making sure the input order is correct.

That's a good point. There is actually no good reason, just that I got lost in my implementation and didn't think about implementing it this way. I will make the change, since what you suggest seems less confusing for users, and requires less checks.

> 3. Indeed, this will break backwards compatibility, but only in Tudat (?)

Yes, it should not change anything in Tudatpy since the exposure links to the new enum emplacements, but users on the Python side will never know the difference.

> 6. Cool! We're going to have to limit what the P&O students use, or their plots will get too full :) Also, with multiple methods of the same order implemented, we're going to have to go into some more detail on error estimates (from a theoretical point of view) and the concept of integrator stability to explain why there are so many different ones! I think this will make a nice addition to the course, and fill a gap that has been there until now

Indeed, we can add so many new integrators, that it can become tricky to justify the use of one over another (and testing all of them will become too cumbersome I guess). I am currently facing this problem with my thesis, I am still busy setting up a "best" integrator for my benchmarks.

About the unit tests: no, I do not have any plan so far. I think I will implement only tests of the orders in a first time, as you suggested. One test that I manually made was also to print all of the Butcher tableaus to make sure they correspond to the documentation I had. It's a trivial test, but it did make me fix two typos, so still quite useful.

About the new integrators you linked: I will try to implement most of them, starting with RK10(8), RK12(10) and RK14(12). Hopefully these will not just result in propagating more numerical error due to precision limits 😬

I will start some of all of these todos this week, but will properly get on them next week :)

**Comment by gaffarelj** [comment]
I also found values for the RKF8(9) coefficients (with 17 stages), in table XII of [this NTRS report](https://ntrs.nasa.gov/api/citations/19680027281/downloads/19680027281.pdf), page 77. 

About the methods such as the one linked [here](https://sce.uhcl.edu/rungekutta/rk108.txt), it should contain two rows for the coefficients of the b matrix (in tudat), since it is an embedded method. However, I only see one row, so I'm a little confused.

**Comment by DominicDirkx** [comment]
Hi Jeremie, by all means implement thee RKF8(9) as well. The RK-Nystrom integrators are not yet compatible with Tudat: these methods only work to integrate second order differential equation. At the moment, Tudat by definition only does 1st order differential equations.

The high-order RK methods I linked to don't have the second row of the Butcher tableau in them explicitly. But, they do have (at the end of the file) a formulation for the LTE (example is for RK10(8):

 _The estimate of the local truncation error is  (1/360)  h ( f(t1,x1)-f(t15,x15) )_

Looking at the [paper](http://www.dynamicpublishers.com/Neural/NPSC2012/NPSC-20-3-4/14-NPSC-32-SA-25.pdf) where the method by which these schemes are defined is described, I see the following:

![image](https://user-images.githubusercontent.com/5039068/163584173-187f246a-29c1-4e29-bc05-a5d5ef5ce1d7.png)

this corresponds to what is in the file for the coefficients of RK12(10). Based on these statements, I'm sure it's possible to derive the second row of b coefficients to formulate the method consistently with the other ones. I'd be happy to derive these coefficients. Even though the error estimate would be much more efficiently estimated by directly implementing the small equations in the files, I suggest that we stick with the standard implementation for now. Doing the extra operations will mostly likely be of almost negligible impact compared to the many state derivative evaluations that need to be done.

**Comment by gaffarelj** [comment]
New update:

1. I have changed the fixed step rk integrator to take lower/higher as input for the order to use (instead of an int representing the order) with b3e862c664139b9980c226cecd86815c3a9e9257.
2. I have added 7 new coefficient sets that can be used for variable step integration (including the one we discussed above), with 95d4c7e79c482af7c345a8b6c911cdeea0eba998.

Changes have been made to tudatpy accordingly, and the new docstrings are in progress.

**Comment by gaffarelj** [comment]
Final (?) update:

1. I fixed the RKF 1(2), Heun(Euler), RKF 8(9) and RKF 14(12) coefficients with 030005a2601150f4fe967c0bab0bf53ae40ed52c. Some orders were switched, and coefficients slightly incorrect.
2. The unit tests were updated with baf4e2a1f45eaf6c037e9a7c38736811ee9d298c to use the new `CoefficientSets` enum that is now out of the `RungeKuttaCoefficients` struct.
3. Running all the unit tests, the following do not pass:
    a. test_orbit_determination_HybridArcStateEstimation
    b. test_propagators_CowellStateDerivative
    c. test_propagators_GaussStateDerivative
    d. test_propagators_GlobalFrameOrigin
    e. test_propagators_VariationalEquations
   The test log can be found here: [LastTest.log](https://github.com/tudat-team/tudat/files/8629399/LastTest.log). Since all of the integrator tests pass (and taking a look at the output), I assume that this is not something caused by me. Could you confirm @DominicDirkx ?
4. I ran an analysis of the number of function evaluations vs error in position for a satellite in a Keplerian orbit with e=0.1, with all of the new coefficient sets, for fixed step intergation. The plots resulting from this, and a CSV file that reports on the slope of the plotted lines, indicating wether the integrator orders are correct or not, can be found here: [variable (lower).pdf](https://github.com/tudat-team/tudat/files/8629410/variable.lower.pdf), [fixed.pdf](https://github.com/tudat-team/tudat/files/8629411/fixed.pdf), [variable (higher).pdf](https://github.com/tudat-team/tudat/files/8629412/variable.higher.pdf), [slopes.csv](https://github.com/tudat-team/tudat/files/8629413/slopes.csv).
5. Looking into the results of point 4, all integrators now seem correct, except the RKF 4(5) and RKF 5(6) which are slightly off. I manually checked that all the coefficients were correct, and they are. However, looking at the Wikipedia page for RKF 4(5), I notice that 3 different Butcher tableaus have been derived (see [here](https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta%E2%80%93Fehlberg_method#Implementing_an_RK4(5)_Algorithm)). I didn't change the coefficients for RKF 4(5) and RKF 5(6), but I left an alternative set of coefficients commented out for the RKF 4(5), which give slightly different results than the currently implemented coefficients, but give a better slope on the test plot from point 4 (the new coefficients give a slope of 4.6(5), the current one a slope of 5(5.5)).

@DominicDirkx if you confirm point 3 (and don't have an issue with anything else), I will open pull requests to merge the `feature/fixed_step_integrators` branch of `tudat`, `tudatpy`, and `tudat-multidoc`.

**Comment by DominicDirkx** [comment]
Thanks, I'm looking forward to getting this into the repo! I will make time in the coming week to dive into the code, but I don't expect any major comments, as I have been following the development. One thing that we should resolve is the unit test failures. From the looks of it, the failures are in numerical integration results where the error _just_ exceeds the permitted tolerance. I'll see if these errors also occur on my machine. Did you get these before you made your modifications? It is possible that you introduced changes that lead to these failures, but my hypothesis is the following if that is the case:

- The cases where there is a failure are ones using the RK4 integrator
- The specific implementation of the RK4 integrator is now slightly different (different order of operations, entries in Butcher tableu defined to higher number of digits), leading to numerically slightly different resuls

Let me know if the errors that are now happening are new, or if you also got them before (which has been known to happen with certain OS/compilers)

All in all: great work, excellent addition to Tudat. In fact, the change is so big that I'll have to modify my lecture material for NA. In fact, the change is so big that I _can_ now modify the lecture material, and discuss/highlight various points much better than before.

**Comment by DominicDirkx** [comment]
I ran all the tests locally, and have made a few small changes, and the unit tests now pass. All the changes cases that failed (and the fixes) are consistent with a change in integrator rounding error. For instance, in one case the errors went away by _increasing_ the time step. In another, we were comparing errors at the mm level on a solar system scale.

I've pushed the changes. I'll have a good look through the code later, but I think that we can merge this soon.
